ipa-core/src/cli/crypto/hybrid_encrypt.rs

```diff
@@ -71,28 +89,199 @@ impl HybridEncryptArgs {
+            let file = OpenOptions::new()
                 .write(true)
                 .create_new(true)
-                .open(self.output_dir.join(&output_filename))
-                .unwrap_or_else(|e| panic!("unable write to {}. {}", &output_filename, e));
-
-            for share in shares {
-                let output = share
-                    .encrypt(DEFAULT_KEY_ID, key_registry, &mut rng)
-                    .unwrap();
-                let hex_output = hex::encode(&output);
-                writeln!(writer, "{hex_output}")?;
-            }
+                .open(output_dir.join(&output_filename))
+                .unwrap_or_else(|e| panic!("unable write to {:?}. {}", &output_filename, e));
+
+            FileWriteWorker::new(file)
+        });
+        let encryptor_pool = EncryptorPool::with_worker_threads(
+            num_cpus::get(),
+            workers.each_ref().map(|x| x.sender.clone()),
+            key_registries,
+        );
+
+        Self {
+            encryptor_pool,
+            workers: Some(workers),
+        }
+    }
+
+    pub fn submit(
+        &mut self,
+        report_id: usize,
+        shares: [HybridReport<BreakdownKey, TriggerValue>; 3],
+    ) -> UnitResult {
+        for (i, share) in shares.into_iter().enumerate() {
+            self.encryptor_pool.encrypt_share((report_id, i, share))?;
         }
 
         Ok(())
     }
+
+    pub fn join(mut self) -> UnitResult {
+        self.encryptor_pool.stop()?;
+        self.workers
+            .take()
+            .unwrap()
+            .map(|worker| {
+                let FileWriteWorker { handle, sender } = worker;
+                drop(sender);
+                handle.join().unwrap()
+            })
+            .into_iter()
+            .collect()
+    }
+}
+
+/// This takes a file and writes all encrypted reports to it,
+/// ensuring the same total order based on `report_id`. Report id is
+/// just the index of file input row that guarantees consistency
+/// of shares written across 3 files
+struct FileWriteWorker {
+    sender: Sender<FileWorkerInput>,
+    handle: JoinHandle<UnitResult>,
+}
+
+impl FileWriteWorker {
+    pub fn new(file: File) -> Self {
+        let (tx, rx) = std::sync::mpsc::channel();
+        Self {
+            sender: tx,
+            handle: thread::spawn(move || {
+                fn write_report<W: Write>(writer: &mut W, report: &[u8]) -> Result<(), BoxError> {
+                    let hex_output = hex::encode(report);
+                    writeln!(writer, "{hex_output}")?;
+                    Ok(())
+                }
+
+                // write low watermark. All reports below this line have been written
+                let mut lw = 0;
+                let mut pending_reports = BTreeMap::new();
+
+                // Buffered writes should improve IO, but it is likely not the bottleneck here.
+                let mut writer = BufWriter::new(file);
+                for (report_id, report) in rx {
```

Comment:
is `rx` different here than above? above it also had a `helper_id`.

Reply:
helper_id is only needed for encryption pool to tell it which file needs to receive encrypted blob. file workers don't need that information because they are bound to a file already